Create rule S7475: Types of unused record components should be removed from pattern matching (SONARJAVA-5432) (#4943)
SonarSource/rspec: rules/S7475/java/rule.adoc

== Why is this an issue?

When using pattern matching on records, matching is done against the canonical constructor of the record.

> This implies listing all the components in the canonical constructor even if there is no use for all of them.

Reviewer: ```suggestion
This implies listing all the components in the canonical constructor even if some are unused.
```

To make the intent of not using the component clear, Java 22 introduced the unnamed variable pattern `_`.

> Because we can only pattern match against the canonical constructor, there is no need to disambiguate.

Reviewer: ```suggestion
Because we can only pattern match against the canonical constructor, there is no need to disambiguate by specifying the types of its parameters.
```

> Therefore, listing the type of unused variables in pattern matching brings no additional value.

Reviewer: ```suggestion
Therefore, the type of unused variables in pattern matching should be omitted, as it does not bring additional value.
```

== How to fix it
Remove the type of the unused component.

=== Code examples

==== Noncompliant code example

[source,java,diff-id=1,diff-type=noncompliant]
----
record Guest(String name, String email, String phoneNumber) {}

String greet(Object o) {
    if (o instanceof Guest(String name, String _, String _)) { // Noncompliant

>         return "Hello "+ name, "!";

Reviewer 2: `name + "!";`

    }
    return "Hello!";
}

String switchToGreet(Object o) {
    return switch (o) {
        case Guest(String name, String _, String _) -> "Hello "+ name + "!"; // Noncompliant
        default -> "Hello!";
    };
}
----

==== Compliant solution

[source,java,diff-id=1,diff-type=compliant]
----
record Guest(String name, String birthPlace) {}

String greet(Object o) {

>     if (o instanceof Guest(String name, _, _)) {

Reviewer 2: `Guest` has only two fields in this example, but there are 3 in the pattern

>         return "Hello "+ name, "!";

Reviewer 2: `,` --> `+`
